src/fabric.c

```diff
@@ -80,15 +83,25 @@ static int fi_register_provider(struct fi_provider *provider)
 
 	prov = fi_getprov(provider->name);
 	if (prov) {
-		/* If we have two versions of the same provider,
-		 * keep the most recent
+		/* If this provider is older than an already-loaded
+		 * provider of the same name, then discard this one.
 		 */
 		if (FI_VERSION_GE(prov->provider->version, provider->version)) {
 			ret = -FI_EALREADY;
 			goto cleanup;
 		}
 
+		/* This provider is newer than an already-loaded
+		 * provider of the same name, so discard the
+		 * already-loaded one.
+		 */
 		prov->provider->cleanup();
+#ifdef HAVE_LIBDL
+		if (prov->dlhandle)
+			dlclose(prov->dlhandle);
+#endif
```

Comment:
I intentionally didn't put #ifdef protection around defining the prov->dlhandle field (up at new line 56) because that would mean that every single use of prov->dlhandle will need to be surrounded by #ifdef (i.e., add #ifdef in a few more places), and I thought people might hate that.

You're right that at run time, prov->dlhandle (and the dlhandle parameter) will only be non-NULL for DLs.  However, if HAVE_LIBDL is false, then libdl is not present, and attempting to find the dlopen() and dlclose() symbols will fail.  Hence, we have to protect all uses of dlopen() and dlclose() inside #ifdef HAVE_LIBDL so that they won't even be compiled in if configure determines that libdl is not present.

Finally, I note that I could have done this:

```c
    if (dlhandle) {
#ifdef HAVE_LIBDL
        dlclose(dlhandle);
#endif
    }
```

but that seemed a bit ugly, so I enclosed the whole if-conditional-and-block inside the #ifdef.